oak_functions/sdk/oak_functions/src/lib.rs

```diff
@@ -0,0 +1,141 @@
+
+//! SDK functionality that provides idiomatic Rust wrappers around the
+//! underlying Oak functions platform functionality.
+
+use log::{debug, error};
+use oak_functions_abi::OakStatus;
+
+/// Reads and returns the user request.
+///
+/// This function is idempotent. Multiple call to this function all return the same value.
+pub fn read_request_body() -> Result<Vec<u8>, OakStatus> {
+    let mut buf = Vec::with_capacity(1024);
+    read_request(&mut buf)?;
+    Ok(buf)
+}
+
+/// Reads the user request into the buffer.
+/// If the buffer does not have enough capacity, it will be reallocated with extra space so that it
+/// can hold the entire request.
+fn read_request(buf: &mut Vec<u8>) -> Result<(), OakStatus> {
+    // Try reading the request twice: first with provided vectors, making
+    // use of their available capacity, then with vectors whose capacity has
+    // been extended to meet size requirements.
+    for resized in &[false, true] {
+        let mut actual_size: u32 = 0;
+        let status_code = unsafe {
+            oak_functions_abi::read_request(buf.as_mut_ptr(), buf.capacity(), &mut actual_size)
+        };
+
+        let status = OakStatus::from_i32(status_code as i32);
+        match status {
+            Some(status) => match status {
+                OakStatus::Ok => {
+                    unsafe {
+                        // The read operation succeeded, and overwrote some fraction of the vector's
+                        // available capacity with returned data (possibly zero). As the data is
+                        // already present in the vector, set its length to match what's available.
+                        buf.set_len(actual_size as usize);
+                    }
+                    return Ok(());
+                }
+                OakStatus::ErrBufferTooSmall if !(*resized) => {
+                    // Extend the vector to be large enough for the message
+                    debug!(
+                        "Got space for {} bytes, need {}",
+                        buf.capacity(),
+                        actual_size
+                    );
+                    if (actual_size as usize) > buf.capacity() {
+                        let extra = (actual_size as usize) - buf.len();
+                        buf.reserve(extra);
+                    }
+
+                    // Try again with a buffer resized to cope with expected size of data.
+                    continue;
+                }
+                status => {
+                    return Err(status);
+                }
+            },
+            None => return Err(OakStatus::ErrInternal),
+        }
+    }
+    error!("unreachable code reached");
+    Err(OakStatus::ErrInternal)
+}
+
+/// Write the response body.
+///
+/// Multiple calls to this function will replace the earlier responses. Only the last response that
+/// is written will be kept and returned to the user.
+pub fn write_response_body(response_body: &str) -> Result<(), OakStatus> {
```

Comment:
I don't think we should have two versions, `&[u8]` should be sufficient, in general the body is not a valid UTF8 string.